3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,253 @@
+   * @private
+   */
+  flushQueues_() {
+    Object.entries(this.creativeMessageRouters_).forEach(
+      entry => {
+        try {
+          const creativeMessageRouter = entry[1];
+          creativeMessageRouter.sendQueuedMessagesToListener();
+        } catch (e) {
+          dev().error(TAG_, 'Caught exception executing listener for ' +
+            entry[0] + ': ' + e.message);
+        }
+      });
+  }
+}
+
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
+   * @param {!string} creativeId
+   */
+  constructor(parent, creativeId) {
+    /** @private {!AmpAnalytics3pMessageRouter} */
+    this.parent_ = parent;
+
+    /**
+     * @private {string}
+     */
+    this.creativeId_ = creativeId;
+
+    /**
+     * @private {string}
+     */
+    this.extraData_ = null;
+
+    /**
+     * private {!Array<ampAnalytics3pEvent>}
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * private {function(Array<ampAnalytics3pEvent>)=}
+     */
+    this.eventListener_ = null;
+
+    if (this.parent_.getWindow().onNewAmpAnalyticsInstance) {
+      this.parent_.getWindow().onNewAmpAnalyticsInstance(this);
+    } else {
+      dev().error(TAG_, 'Vendor page must implement onNewAmpAnalyticsInstance' +
+        ' prior to loading library script.');
+    }
+  }
+
+  /**
+   * Registers a callback function to be called when AMP Analytics events occur.
+   * There may only be one listener. If another function has previously been
+   * registered as a listener, it will no longer receive events.
+   * @param {!function(!Array<ampAnalytics3pEvent>)} listener A function
+   * that takes an array of event strings, and does something with them.
+   */
+  registerAmpAnalytics3pEventsListener(listener) {
+    this.eventListener_ = listener;
+  }
+
+  /**
+   * Receives a message from a creative for the cross-domain iframe
+   * @param {ampAnalytics3pNewCreative} message The message that was received
+   */
+  receiveNewCreativeMessage(message) {
+    this.extraData_ = message[AMP_ANALYTICS_3P_MESSAGE_TYPE.NEW];
+  }
+
+  /**
+   * Receives a message from a creative for the cross-domain iframe
+   * @param {ampAnalytics3pEvent} message The message that was received
+   */
+  receiveEventMessage(message) {
+    this.receivedMessagesQueue_.push(message);
+  }
+
+  /**
+   * If an event listener has been created on the third-party vendor's
+   * metrics-collection page, this method passes the queued events to that
+   * listener.
+   * If there is no event listener on the page yet, event messages will remain
+   * in the queue.
+   */
+  sendQueuedMessagesToListener() {
+    if (!this.receivedMessagesQueue_ || !this.eventListener_) {
+      return;
+    }
+    this.eventListener_(this.receivedMessagesQueue_);
```

Comment:
try/catch should be here to ensure that error does not prevent queue being flushed